src/main/scala/loamstream/model/execute/ExecutionState.scala

```diff
@@ -0,0 +1,302 @@
+          debug(s"Restarting $job ? YES (job has run $runCount times, max is $maxRunsPerJob)")
+          
+          _.markAsRunnable 
+        }
+        else {_.finishWith(status) }
+      }
+      
+      jobStates(jobIndex) = JobState(job, transition(cell))
+      
+      if(isTerminalFailure) {
+        cancelSuccessors(jobStates)(job)
+      }
+    }
+  }
+  
+  /**
+   * Cancel all the jobs that depend on this job, and the ones that depend on them, etc, by
+   * marking them as CouldNotStart.
+   */
+  private[execute] def cancelSuccessors(fields: JobStateList)(failedJob: LJob): Unit = {
+    TimeUtils.time(s"Cancelling successors for failed job with id ${failedJob.id}", debug(_)) {
+      val successors = ExecuterHelpers.flattenTree(Set(failedJob), _.successors).toSet - failedJob
+      
+      markAs(successors.iterator.map(_.job), JobStatus.CouldNotStart)
+    }
+  }
+}
+
+object ExecutionState {
+  /**
+   * The data structure that tracks the state of all the jobs being executed by LoamStream on a given run.
+   * The data is represented as a sequence of JobStates (Job, ExecutionCell) tuples.  An array is used for a
+   * few reasons, mostly related to performance.  Initally, I tried a Map[LJob, ExecutionCell], but while this
+   * allowed fast lookups of the state of a job, iterating through the Map to find runnable jobs (something LS
+   * does very frequently) was too slow. Updates to lots of jobs in the immutable Map (the norm when running 
+   * large pipelines) also produced a large amount of GC pressure.  
+   *  
+   * A mutable array addresses both of these: iterating over an array is as fast as it gets on the JVM, and only
+   * making new ExecutionCells on job state changes instead of new Maps substantially reduced the amount of 
+   * allocations and GC activity.  Together, these factors increased performance a lot.
+   * 
+   * Looking up JobStates by jobs, and finding the subset of JobStates for a given bunch of jobs is made fast 
+   * (enough) by pairing a JobStateList with an index, a Map[LJob, Int], mapping jobs to their positions in a
+   * JobStateList.   
+   */
+  private final class JobStateList(byJob: Array[JobState]) extends Iterable[JobState] {
+    def snapshot: JobStateList = new JobStateList(copyOf(byJob))
+    
+    override def size: Int = byJob.length
+    
+    override def iterator: Iterator[JobState] = byJob.iterator
+    
+    def apply(i: Int): JobState = byJob(i)
+    
+    def update(i: Int, jobState: JobState): Unit = byJob.update(i, jobState)
+    
+    def cellFor(index: Map[LJob, Int])(job: LJob): ExecutionCell = byJob(index(job)).cell
+  }
+  
+  private def copyOf(a: Array[JobState]): Array[JobState] = {
+    val result: Array[JobState] = Array.ofDim(a.length)
+    
+    a.copyToArray(result)
+    
+    result
+  }
+  
+  /**
+   * Basically a tuple of a job and its current status, as represented by an ExecutionCell
+   */
+  final case class JobState(job: LJob, cell: ExecutionCell) {
```

Comment:
I think if I had a "major" comment about the `ExecutionCell` and `ExecutionState` that I'd push for in the interest of simpler code (subjective), it'd be this:

* Get rid of `ExecutionCell`.

Try and do as much as possible with just `JobStatus`. There's likely a 1-2 "small" things that `ExecutionCell` does on top of `JobStatus` that make it valuable, but I feel like those should either be horizontal (in `ExecutionState`) or vertical (as a subclass of `JobStatus`) rather than a parallel implementation that appears to duplicate a lot of functionality/querying.

Possible examples (pseudo-code):

```scala
// Horizontal...
//
// Uses JobStatus to get the final "finished" status of complete vs. failure
//
// Issue: managing multiple sets where the jobs in them are mutually exclusive (can cause bugs)
class ExecutionState {
  var runnableJobs: Seq[Job]
  var runningJobs: Seq[Job]
  var finishedJobs: Seq[Job]
}

// Vertical...
//
// Keeps extended JobStatus "private" to the ExecutionState, keeps the name, etc.
class ExecutionState {
  val jobStatuses = Seq[ExecutionState.JobStatus]
}

object ExecutionState {
  class JobStatus extends base.JobStatus {
    // markAs, ...
  }
}
```

Reply:
I like keeping `JobStatus` basically an "enum" with broad applicability for now.  There have been enough cases already where I've needed to build extra behavior on top of or around it that I'm wary of baking behavior specific to a particular use into that class. 

I merged `ExecutionState.JobState` and `ExecutionCell` into a new class `JobExecutionState`, which makes the current code more or less the "vertical" option.